libtenzir/builtins/operators/chart2.cpp

```diff
@@ -0,0 +1,794 @@
+      y,
+      [&](ast::record& rec) -> failure_or<void> {
+        for (auto& i : rec.items) {
+          auto* field = try_as<ast::record::field>(i);
+          if (not field) {
+            diagnostic::error("cannot use `...` here").primary(y).emit(ctx);
+            return failure::promise();
+          }
+          const auto loc = field->expr.get_location();
+          TRY(match(
+            field->expr,
+            [&](ast::function_call& call) -> failure_or<void> {
+              args.y[field->name.name] = std::move(call);
+              return {};
+            },
+            [&](auto& expr) -> failure_or<void> {
+              if (args.res) {
+                diagnostic::error("an aggregation is required")
+                  .primary(field->expr)
+                  .emit(ctx);
+                return failure::promise();
+              }
+              auto result
+                = ast::function_call{entity, {std::move(expr)}, loc, false};
+              TENZIR_ASSERT(resolve_entities(result, ctx));
+              args.y[field->name.name] = std::move(result);
+              return {};
+            }));
+        }
+        return {};
+      },
+      [&](ast::function_call& call) -> failure_or<void> {
+        args.y["y"] = std::move(call);
+        return {};
+      },
+      [&](auto&) -> failure_or<void> {
+        if (args.res) {
+          diagnostic::error("an aggregation is required").primary(y).emit(ctx);
+          return failure::promise();
+        }
+        const auto loc = y.get_location();
+        auto result = ast::function_call{entity, {std::move(y)}, loc, false};
+        TENZIR_ASSERT(resolve_entities(result, ctx));
+        args.y["y"] = std::move(result);
+        return {};
+      });
+  }
+
+  auto handle_limit(const chart_args& args, ast::binary_op op,
+                    const located<data>& limit, diagnostic_handler& dh) const
+    -> failure_or<xlimit> {
+    const auto& loc = limit.source;
+    auto result = match(
+      limit.inner,
+      [&](const caf::none_t&) -> failure_or<ast::constant> {
+        diagnostic::error("limit cannot be `null`").primary(limit).emit(dh);
+        return failure::promise();
+      },
+      [&](const pattern&) -> failure_or<ast::constant> {
+        diagnostic::error("limit cannot be a pattern").primary(limit).emit(dh);
+        return failure::promise();
+      },
+      [&](const duration& d) -> failure_or<ast::constant> {
+        if (args.res) {
+          const auto val = d.count();
+          const auto count = std::abs(args.res->inner.count());
+          const auto rem = std::abs(val % count);
+          if (rem) {
+            const auto ceil = val >= 0 ? count - rem : rem;
+            const auto floor = val >= 0 ? -rem : rem - count;
+            return ast::constant{
+              duration{val + (op == ast::binary_op::leq ? floor : ceil)},
+              loc,
+            };
+          }
+        }
+        return ast::constant{d, loc};
+      },
```

Comment:
Related to that: Now that the min/max arguments are done visually, would it make sense to only use the floored/ceiled value for predicate pushdown, but still limit the chart visually to the original min/max? At least for line and area charts that would probably be what you'd want. For bar charts, I'm not so sure what to do.

Reply:
I think that's dishonest from the user's perspective, might cause confusion.

Reply:
And why do you think so? I would say we'd follow the user's request to the letter when abiding by their `x_min` as specified. Note I'm not saying that we _have_ to do it, just help me understand why you think it would be a problem.

Reply:
If we do so, I would push down the original values. Pushing and aggregating over a different set of data yet still showing the originally specified limits seems confusing to me.

Reply:
The motivation for pushing down the floored values was because we said that it's better to have extra data in the chart that the user did not request than to not show data for the given limit at all, right? Now, this still holds true, and we don't want to not show data that the user requested. But my proposal above shows everything that is requested, plus it even follows the specified `x_min` more closely. That we are internally computing beyond the specified interval seems like an implementation detail to me. But the user doesn't think about that, all they care about is that their chart starts at `x_min`.

Reply:
I think it's easier if we huddle over this?